### PR TITLE
[3468] Make a new EventHandler to support a sequential event processing

### PR DESCRIPTION
**Pull request by kanat**
Closes https://github.com/GetStream/stream-chat-android/issues/3468.

### 🎯 Goal

The goal is to provide another implementation of `EventHandler` to process events sequentially to avoid inconsistency in database & states.

### 🛠 Implementation details

`EventHandlerSequential` is currently enabled for `DEBUG` builds only as per agreement with @adasiewiczr to let us test it first before releasing.
```kotlin
return when (BuildConfig.DEBUG) {
    true -> EventHandlerSequential(
        scope = scope,
        recoveryEnabled = true,
        subscribeForEvents = { listener -> client.subscribe(listener) },
        logicRegistry = logicRegistry,
        stateRegistry = stateRegistry,
        mutableGlobalState = mutableGlobalState,
        repos = repos,
        syncManager = syncManager,
    )
    else -> EventHandlerImpl(
        scope = scope,
        recoveryEnabled = true,
        client = client,
        logic = logicRegistry,
        state = stateRegistry,
        mutableGlobalState = mutableGlobalState,
        repos = repos,
        syncManager = syncManager,
    )
}
```

### 🧪 Testing

A full regression is required.

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [ ] ~~Comparison screenshots added for visual changes~~
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_

### CODE REVIEW

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventBatchUpdate.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventBatchUpdate.kt:51]: Nice one

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerImpl.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerImpl.kt:104]: Maybe let's call it `Chat:EventHandlerOld`

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerImpl.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerImpl.kt:136]: What do you think about decreasing log level here?

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerImpl.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerImpl.kt:151]: and here

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandler.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandler.kt:49]: Should we change the parameter name to `events`?
Also we should add KDocs

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:165]: Should we also introduce `deinitialize` just to make sure `currentUserId ` is cleared and `initJob` joined? 🤔

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:111]:
```suggestion
private const val TAG = "Chat: EventHandlerSequential"
```

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/utils/ResultUtils.kt
Comment (adasiewiczr) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/utils/ResultUtils.kt:24]: Let's add KDocs to both extension functions

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:795]: Just out of curiosity - why 300?

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:724]:
```suggestion
            override fun dispose() = Unit
```
this should remove Detekt warning

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:736]: Do we need this getter? Cannot we just use `isFromHistorySync` parameter and/or change it's name

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:731]: Let's add a short KDoc (where it's used)

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:739]: Should we extract this to a separate file?
Some KDocs will be helpful as well

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:748]: Should we crash or at least log an error in another case? 🤔

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:760]: 🤔 It's an edge case but what will happen if we are going to constantly receive events every 200ms?

**Review by adasiewiczr — COMMENTED** [review]
Very good job!
I left some small comments and questions but looks amazing overall + works as expected

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (kanat) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:736]: Yup we can omit getter here 👍
I've added that opposite param for the sake of readability.

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (kanat) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:748]: We should not cause it collects just those 2 events for now. Other events are ignored.

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (kanat) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:760]: Good thinking.
We should probably limit the max size of the batch.

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (adasiewiczr) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:748]: I would still add a log just to make sure it works as expected if someone decides to change the code

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (kanat) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt]: Based on max gap of 200ms I've seen in logs between 2 events

**Review by adasiewiczr — APPROVED** [review]
LGTM but it's a big chagne and let's wait for a second review. @amitkma @leandroBorgesFerreira 

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (leandroBorgesFerreira) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:463]: I think some code here is duplicated in `EventHandlerImpl`. It would be nice to merge both of them because we will have to remember to always fix in both places if we have to change the code here. It can be challenging to keep both implementations the same as time passes.

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/batch/SocketEventCollector.kt
Comment (leandroBorgesFerreira) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/batch/SocketEventCollector.kt:51]: If I understand this correctly, we are always waiting 300ms to batch and fire events. In the case of a single event or a sparse array of events, this will slow performance (it doesn't trigger events right away). Couldn't we keep this logic, but handle this with some kind of backpressure instead of always batching events by time?

**Review by leandroBorgesFerreira — COMMENTED** [review]
2 comments

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt
Comment (kanat) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt:463]: Hey @leandroBorgesFerreira,

That's because of the agreement to keep `EventHandlerImpl` to avoid breaking anything and introduce a new `EventHandler` to replace the existing one if no issues are found.

**Pull request merged**
